src/trace.rs

```diff
@@ -0,0 +1,331 @@
+        if let Some(node) = self.unobserved.pop() {
+            // scope in for write access to node
+            {
+                let mut node_w = node.0.write().expect("poisoned TraceNode");
+                if let Some(max_term_size) = self.max_term_size {
+                    if node_w.term.size() > max_term_size {
+                        node_w.state = TraceState::TooBig;
+                        return None;
+                    }
+                }
+                match self.trs.rewrite(&node_w.term) {
+                    Some(ref rewrites) if !rewrites.is_empty() => {
+                        let term_selection_p = -(rewrites.len() as f64).ln();
+                        for term in rewrites {
+                            let new_p =
+                                node_w.log_p + (1.0 - self.p_observe).ln() - term_selection_p;
+                            let new_node = self.new_node(
+                                term.clone(),
+                                Some(&node),
+                                TraceState::Unobserved,
+                                new_p,
+                            );
+                            node_w.children.push(new_node);
+                        }
+                        node_w.log_p += self.p_observe.ln()
+                    }
+                    _ => node_w.state = TraceState::Normal,
+                }
+            }
+            Some(node)
+        } else {
+            None
+        }
+    }
+    /// Run the `Trace` until a fruitful step is reached.
+    pub fn run(&mut self, mut max_steps: usize) {
+        while let None = self.step() {
+            max_steps -= 1;
+            if max_steps == 0 {
+                break;
+            }
+        }
+    }
+    /// Run the `Trace` until a fruitful step is reached and return the leaf terms.
+    pub fn rewrite(&mut self, max_steps: usize, states: &[TraceState]) -> Vec<Term> {
+        self.run(max_steps);
+        self.root.leaf_terms(states)
+    }
+    /// An estimate of the probability that `self` rewrite to `term`.
+    pub fn rewrites_to(&mut self, max_steps: usize, term: &Term) -> f64 {
+        // NOTE: we only use tree equality and don't consider tree edit distance
+        self.run(max_steps);
+        let lps = self.root.accumulate(|n| {
+            let n_r = &n.0.read().expect("poisoned TraceNode");
+            Term::alpha(term, &n_r.term).map(|_| n_r.log_p)
+        });
+        if lps.is_empty() {
+            f64::NEG_INFINITY
+        } else {
+            logsumexp(&lps)
+        }
+    }
+}
+
+/// A single node in a Trace.
+#[derive(Debug, Clone)]
+struct TraceNodeStore {
+    term: Term,
+    state: TraceState,
+    log_p: f64,
+    depth: usize,
+    parent: Option<Weak<RwLock<TraceNodeStore>>>,
+    children: Vec<TraceNode>,
+}
+
+/// The state of a Node in a Trace
+#[derive(Debug, Clone, Copy, PartialEq, Eq, Hash)]
+pub enum TraceState {
+    Start,
+    Normal,
+    Unobserved,
+    TooBig,
+}
```

Comment:
As you'll see in my branch, I realized that we don't need `Start`, but we do need something like `Rewritten` for nodes that have been observed but aren't too big or normal forms.